changes/changes_test.go

```diff
@@ -0,0 +1,550 @@
+package changes
+
+import (
+	"context"
+	"encoding/json"
+	// "fmt"
+	"io/ioutil"
+	"os"
+	"sort"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	golog "github.com/ipfs/go-log"
+	"github.com/qri-io/dataset"
+	"github.com/qri-io/dataset/tabular"
+	"github.com/qri-io/qfs"
+	"github.com/qri-io/qri/base"
+	"github.com/qri-io/qri/base/dsfs"
+	"github.com/qri-io/qri/dsref"
+	"github.com/qri-io/qri/repo"
+	repotest "github.com/qri-io/qri/repo/test"
+	"github.com/qri-io/qri/stats"
+)
+
+func newTestService(t *testing.T, r repo.Repo, workDir string) *Service {
+	// debug logging
+	golog.SetLogLevel("changes", "debug")
+
+	cache, err := stats.NewLocalCache(workDir, 1000<<8)
+	if err != nil {
+		t.Fatal(err)
+	}
+	statsSvc := stats.New(cache)
+	loader := base.NewLocalDatasetLoader(r.Filesystem())
+
+	return New(loader, statsSvc)
+}
+
+func updateDataset(t *testing.T, r repo.Repo, ds *dataset.Dataset, newBody string) dsref.Ref {
+	ctx := context.Background()
+	currRef := dsref.ConvertDatasetToVersionInfo(ds).SimpleRef()
+
+	ds.SetBodyFile(qfs.NewMemfileBytes("body.csv", []byte(newBody)))
+	ds.PreviousPath = currRef.Path
+
+	res, err := base.CreateDataset(ctx, r, r.Filesystem().DefaultWriteFS(), ds, nil, dsfs.SaveSwitches{Pin: true, ShouldRender: true})
+	if err != nil {
+		t.Fatal(err.Error())
+	}
+	return dsref.ConvertDatasetToVersionInfo(res).SimpleRef()
+}
+
+func TestStatsDiff(t *testing.T) {
+	ctx := context.Background()
+
+	workDir, err := ioutil.TempDir("", "qri_test_changes_service")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(workDir)
+
+	mr, err := repotest.NewTestRepo()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	svc := newTestService(t, mr, workDir)
+
+	ref := dsref.MustParse("peer/cities")
+	if _, err := mr.ResolveRef(ctx, &ref); err != nil {
+		t.Fatal(err)
+	}
+
+	ds, err := dsfs.LoadDataset(ctx, mr.Filesystem(), ref.Path)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err = base.OpenDataset(ctx, mr.Filesystem(), ds); err != nil {
+		t.Fatal(err)
+	}
+
+	ds.Name = "cities"
+	leftDs := *ds
+
+	// alter body file
+	const alteredBodyData = `city,pop,avg_age,in_usa
+toronto,4000000,55.0,false
+new york,850000,44.0,false
+chicago,30000,440.4,false
+chatham,3500,650.25,false
+raleigh,25000,5000.65,false`
+
+	updateDataset(t, mr, ds, alteredBodyData)
+
+	res, err := svc.statsDiff(ctx, &leftDs, ds)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// output order is not strict so we need to enfore it here
+	sort.SliceStable(res.Columns, func(i, j int) bool {
+		return res.Columns[i].Title < res.Columns[j].Title
+	})
+
+	expect := &StatsChangeComponent{
+		Summary: &ChangeReportDeltaComponent{
+			ChangeReportComponent: ChangeReportComponent{
+				Left:  StatsChangeSummaryFields{Entries: 5, Columns: 4, TotalSize: 157},
+				Right: StatsChangeSummaryFields{Entries: 5, Columns: 4, TotalSize: 157}, // is this right?
+			},
+			Delta: StatsChangeSummaryFields{},
```

Comment:
This seems to be empty, think we need to populate the delta on the summary too. Leaving this mostly as a note to myself.